Add new methods to get the number of actions running in a given node with a specific tag.

saturnodigital opened the pull request:
This is useful for cases that we want to know how many
animations with a **specific** tag is running in our target.
For example:

auto tag1Count = this->getNumberOfRunningActionsByTag(kMyTag1);
auto tag2Count = this->getNumberOfRunningActionsByTag(kMyTag2);

While this could be achieved (more or less) with callbacks
to increment the tagCount at start of action and decrement
at the end the proposed API is much more concise and less error
prone.

saturnodigital committed:
Add new methods to get the number of actions running in a
given node with a specific tag.

This is useful for cases that we want to know how many
animations with a **specific** tag is running in our target.
For example:

auto tag1Count = this->getNumberOfRunningActionsByTag(kMyTag1);
auto tag2Count = this->getNumberOfRunningActionsByTag(kMyTag2);

While this could be achieved (more or less) with callbacks
to increment the tagCount at start of action and decrement
at the end the proposed API is much more concise and less error
prone.

minggo commented on cocos/2d/CCActionManager.cpp:
```diff
@@ -398,6 +398,32 @@ ssize_t ActionManager::getNumberOfRunningActionsInTarget(const Node *target) con
+    auto limit = element->actions->num;
+    for(int i = 0; i < limit; ++i)
+    {
+        auto action = (Action *)element->actions->arr[i];
```
it is better to use `static_cast<Action*>` instead

saturnodigital replied:
@minggo I just used the same pattern of the "surrounding" methods.
I agree with the static_assert, but we should change for the other methods too?

minggo replied:
You are appreciated if you can help to fix other methods.

minggo commented on cocos/2d/CCActionManager.h:
```diff
@@ -145,6 +145,24 @@ class CC_DLL ActionManager : public Ref
+     * @see getNumberOfRunningActionsInTarget
+     * @js NA
+     */
+    size_t getNumberOfRunningActionsInTargetByTag(const Node *target, int tag);
```
it is better to add const like 

``` c++
size_t getNumberOfRunningActionsInTargetByTag(const Node *target, int tag) const;
```

minggo replied:
and it is better to use reference instead like 

``` c++
size_t getNumberOfRunningActionsInTargetByTag(const Node &target, int tag) const;
```

or we should check the invalidation of target.

saturnodigital replied:
@minggo I just used the same pattern of the "surrounding" methods.
Actually I didn't understand your comment "it is better to add const like ...", my PR 
is already using ``const Node *`, isn't it?

About the reference, I have no ideas what would be better, we have a lot of comments with this issue:

```
// FIXME: Passing "const O *" instead of "const O&" because HASH_FIND_IT requires the address of a pointer
// and, it is not possible to get the address of a reference
```

But this could go deeper in the engine architecture than I could make a clear statement about.
I leave to you guys decide what is better, regarding this modification.

Anyway is just let me know about that I'll happily make another PR with the changes!

minggo replied:
ok, let's keep the same pattern at this PR

cmzx3444 commented on cocos/2d/CCActionManager.cpp:
```diff
@@ -398,6 +398,32 @@ ssize_t ActionManager::getNumberOfRunningActionsInTarget(const Node *target) con
 
+// FIXME: Passing "const O *" instead of "const O&" because HASH_FIND_IT requires the address of a pointer
+// and, it is not possible to get the address of a reference
+size_t ActionManager::getNumberOfRunningActionsInTargetByTag(const Node *target,
```
Method name can be shorter? like getRunningActionsNumber?

saturnodigital replied:
I have the same feel, but I just  followed the pattern in the code!
The other methods are "overloaded" by adding the "byTag" suffix, so I did the same.

For the other side, I think that is easier to make a mental model with this name since
we are used to the other methods that uses the "byTag" suffix.

minggo replied:
follow the same pattern looks good to me

saturnodigital committed:
Replace the old C casts to static_casts.

As discussed in #16789 issue thread @minggo asked
to change the old style of cast to newer, more safer
static_casts.

saturnodigital committed:
Add the required parenthesis and normalize the spaces.

Normalize all the static_cast<Action *> to static_cast<Action*>
some places we have spaces, other places haven't.

Fix the missing parenthesis.
